lib/scholar/neighbors/graph.ex

```diff
@@ -0,0 +1,182 @@
+defmodule Scholar.Neighbors.Graph do
+  import Nx.Defn
+  require Nx
+  alias Scholar.Neighbors.RandomProjectionForest, as: Forest
+
+  knn_schema = [
+    k: [
+      required: true,
+      type: :pos_integer,
+      doc: "The number of neighbors in the graph."
+    ]
+  ]
+
+  # TODO: Rename this
+  approximate_knn_schema =
+    knn_schema ++
+      [
+        min_leaf_size: [
+          type: :pos_integer,
+          doc: "The minimum number of points in ..."
+        ],
+        num_trees: [
+          type: :pos_integer,
+          doc: "The number of trees in random projection forest."
+        ],
+        num_iters: [
+          type: :non_neg_integer,
+          default: 1,
+          doc: "The number of times to perform neighborhood expansion."
+        ],
+        key: [
+          type: {:custom, Scholar.Options, :key, []},
+          doc: """
+          Used for random number generation in parameter initialization.
+          If the key is not provided, it is set to `Nx.Random.key(System.system_time())`.
+          """
+        ]
+      ]
+
+  @knn_schema NimbleOptions.new!(knn_schema)
+  @approximate_knn_schema NimbleOptions.new!(approximate_knn_schema)
+
+  deftransform exact_brute_force(x, opts) do
+    opts = NimbleOptions.validate!(opts, @knn_schema)
+    exact_brute_force_n(x, opts)
+  end
+
+  defn exact_brute_force_n(x, opts) do
+    k = opts[:k]
+    # TODO: Add other metrics other than Euclidean
+    x
+    |> Nx.new_axis(1)
+    |> Nx.subtract(Nx.new_axis(x, 0))
+    |> Nx.pow(2)
+    |> Nx.sum(axes: [2])
+    |> Nx.put_diagonal(Nx.broadcast(:infinity, {Nx.axis_size(x, 0)}))
+    |> Nx.negate()
+    |> Nx.top_k(k: k)
+    |> elem(1)
+  end
+
+  @doc """
+  Constructs the approximate k-nearest neighbor graph using random projection forest
+  and neighborhood expansion.
+  """
+  deftransform approximate_random_projection_forest_and_expansion(x, opts) do
```

Comment:
So what is the algorithm implemented here? LargeVis? NNExpansion? I think we should aim to implement some particular algorithms and then, if there are shared paths, for example, "expansion" logic, we move that to the forests module?

Reply:
The algorithm works as follows: first it initializes the k-NN graph using random projection forest, then refines it using NN-expansion. The algorithm is similar to the one used in LargeVis (although the idea is pretty natural and I had it even before stumbling upon that paper; something similar is used in other algorithms). The main difference is how the trees are constructed. LargeVis uses Annoy which, at every node, randomly picks two points, separates them with a hyperplane, and then splits the data depending on which side of the hyperplane each point is.

I would keep the expansion in this module or create a separate one for it as expansion itself can give decent results for k-NN graph construction (see e.g. EFANNA paper).

Reply:
> Ideally, we would have NNDescent instead of NN-expansion.

It is ok to have multiple implementations. The only trouble is figuring out how we are going to expose this. Calling it `Graph` is a bit generic. Do we expect all implementations of kNN-graph to be in this module or follow the struct returned by this module? I would say that this is too early to say and, generally speaking, I can easily see us having different knn-graphs in the future.

I would prefer if we could stick with the fit/predict APIs, but you are more knowledgeable than me on what is the fit/predict here. I assume this function is a `fit` and we need to add a `predict` to it? If we do such, we just call it `NNExpansion`?

Reply:
> > Ideally, we would have NNDescent instead of NN-expansion.
> 
> It is ok to have multiple implementations. The only trouble is figuring out how we are going to expose this. Calling it `Graph` is a bit generic. Do we expect all implementations of kNN-graph to be in this module or follow the struct returned by this module? I would say that this is too early to say and, generally speaking, I can easily see us having different knn-graphs in the future.
> 
> I would prefer if we could stick with the fit/predict APIs, but you are more knowledgeable than me on what is the fit/predict here. I assume this function is a `fit` and we need to add a `predict` to it? If we do such, we just call it `NNExpansion`?

Yes, agreed, `Graph` is too generic. Perhaps there should be a module for every k-NN graph construction algorithm. Also, calling this module `NNExpansion` seems like a good idea as long as we mention that we are using random projection forest for initialization. I think we need only `fit` function that takes a tensor and opts and returns a tensor of indices of shape `{n, k}` where `i`-th row contains `k`-nearest neighbors of the `i`-th point. Don't think `predict` is needed here.

Reply:
So to make sure I get it:

1. We have different ways to build graphs

2. NNExpansion is one of such ways and it *may* use RandomForests

3. RandomForests can be used to build graphs but NNExpansion is only one of the ways to do that with RandomForests

So here is the question: can I run NNExpansion with kdtrees? Because if we can, then I think we should make the tree/forest algorithm an argument of the NNExpansion algorithm. This way they can compose.

Reply:
Not sure if the current implementation of k-d trees can construct an approximate k-NN graph, but in principle yes, we could use them as well. Another possibility is to use random initialization, i.e. for every point randomly choose `k` other points as its `k`-nearest neighbors.